### PR TITLE
make do_queue=False fail for Sum and Prod construction

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
`do_queue=False` causes unexpected behaviour with composite operators, especially Prod. For example, if you do `qml.prod(qml.PauliX(0), qml.PauliZ(0), do_queue=True)`, it queues $X0 @ Z0$. But if you do it with `do_queue=False`, it doesn't queue the prod, but it also doesn't de-queue the X then Z (in that order), which is equivalent to $Z0 @ X0$, the reverse of the previous case! To avoid ambiguity, users should consider using `with qml.stop_recording()` instead.

**Description of the Change:**
Make `do_queue=False` raise an error in all places related to composite ops.

**Benefits:**
There is no unexpected queuing behaviour with composite ops.

**Possible Drawbacks:**
Users cannot avoid queuing with a kwarg and will need to use the context instead.

**Related GitHub Issues:**
Found while working on #4011

### CODE REVIEW

File: pennylane/ops/op_math/prod.py [hunk 1]
```diff
@@ -85,20 +83,12 @@ def prod(*ops, do_queue=True, id=None, lazy=True):
...
-        return Prod(*ops, do_queue=do_queue, id=id)
-
-    ops_simp = Prod(
-        *itertools.chain.from_iterable([op if isinstance(op, Prod) else [op] for op in ops]),
-        do_queue=do_queue,
-        id=id,
-    )
-
-    if do_queue:
+    if not lazy:
```
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/ops/op_math/prod.py:90]: we only do extra work when `lazy=False` so this makes sense to me

File: pennylane/ops/op_math/composite.py [hunk 2]
```diff
@@ -35,7 +35,7 @@ class CompositeOp(Operator):
         operands: (tuple[~.operation.Operator]): a tuple of operators which will be combined.
 
     Keyword Args:
-        do_queue (bool): determines if the operator will be queued. Default is True.
+        do_queue (bool): Must be True (default) for CompositeOps, but present for consistency.
```
Comment (trbromley) [inline pennylane/ops/op_math/composite.py:38]:
```suggestion
        do_queue (bool): Must be ``True`` (default) for ``CompositeOps``, but present for consistency.
```

File: pennylane/ops/op_math/prod.py [hunk 3]
```diff
@@ -54,7 +53,7 @@ def prod(*ops, do_queue=True, id=None, lazy=True):
         ops (tuple[~.operation.Operator]): The operators we would like to multiply
 
     Keyword Args:
-        do_queue (bool): determines if the product operator will be queued. Default is True.
+        do_queue (bool): Must be True (default) for prod, but present for consistency.
```
Comment (trbromley) [inline pennylane/ops/op_math/prod.py:56]:
```suggestion
        do_queue (bool): Must be ``True`` (default) for ``prod``, but present for consistency.
```

File: pennylane/ops/op_math/prod.py [hunk 4]
```diff
@@ -109,7 +103,7 @@ class Prod(CompositeOp):
         together.
 
     Keyword Args:
-        do_queue (bool): determines if the product operator will be queued. Default is True.
+        do_queue (bool): Must be True (default) for Prod, but present for consistency.
```
Comment (trbromley) [inline pennylane/ops/op_math/prod.py:106]:
```suggestion
        do_queue (bool): Must be ``True`` (default) for ``Prod``, but present for consistency.
```

File: pennylane/ops/op_math/sum.py [hunk 5]
```diff
@@ -48,8 +48,7 @@ def sum(*summands, do_queue=True, id=None, lazy=True):
         summands (tuple[~.operation.Operator]): the operators we want to sum together.
 
     Keyword Args:
-        do_queue (bool): determines if the sum operator will be queued (currently not supported).
-            Default is True.
+        do_queue (bool): Must be True (default) for sum, but present for consistency.
```
Comment (trbromley) [inline pennylane/ops/op_math/sum.py:51]: Ok I wish I hadn't made the suggestion now :laughing:

Comment (trbromley) [inline pennylane/ops/op_math/sum.py:51]:
```suggestion
        do_queue (bool): Must be ``True`` (default) for ``sum``, but present for consistency.
```

**Review by trbromley — APPROVED** [review]
Thanks @timmysilv!

File: pennylane/ops/op_math/composite.py [hunk 6]
```diff
@@ -55,16 +55,16 @@ def __init__(
         if len(operands) < 2:
             raise ValueError(f"Require at least two operators to combine; got {len(operands)}")
 
+        if not do_queue:
+            raise ValueError(f"do_queue=False is not supported for {self.name}")
```
Comment (trbromley) [inline pennylane/ops/op_math/composite.py:59]: Oh actually one thought, how about we give users a suggested path? Like `stop_recording`?

**Review by mudit2812 — APPROVED** [review]

File: pennylane/ops/op_math/sum.py [hunk 7]
```diff
@@ -106,7 +101,7 @@ class Sum(CompositeOp):
         summands (tuple[~.operation.Operator]): a tuple of operators which will be summed together.
 
     Keyword Args:
-        do_queue (bool): determines if the sum operator will be queued. Default is True.
+        do_queue (bool): Must be True (default) for Sum, but present for consistency.
```
Comment (timmysilv) [inline pennylane/ops/op_math/sum.py:104]:
```suggestion
        do_queue (bool): Must be ``True`` (default) for ``Sum``, but present for consistency.
```

**Comment by timmysilv** [comment]
Frederik is going to deprecate `do_queue` everywhere in a way that makes more sense for deprecation (rather than my hard-fail here), then add in a line or two to make it actually fail like it does in this PR. It makes sense because while we need to fail if it's false, we now also want to warn if it's true, which this PR does not consider so it's written without that consideration. Closing in favour of Frederik's upcoming change

**Pull request closed without merging**